pkg/agent/controllers/agent/agent.go

```diff
@@ -0,0 +1,205 @@
+package agent
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+)
+
+const (
+	handlerResync          = time.Hour * 2
+	submarinerIpamGlobalIp = "submariner.io/globalIp"
+)
+
+var excludeSvcList = map[string]bool{"kubernetes": true, "openshift": true}
+
+func New(spec *LighthouseAgentSpecification, config InformerConfigStruct) (*Controller, error) {
+	exclusionNSMap := map[string]bool{"kube-system": true, "submariner": true, "openshift": true,
+		"submariner-operator": true, "kubefed-operator": true}
+	for _, v := range spec.ExcludeNS {
+		exclusionNSMap[v] = true
+	}
+	agentController := &Controller{
+		kubeClientSet:    config.KubeClientSet,
+		serviceWorkqueue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Services"),
+		servicesSynced:   config.ServiceInformer.Informer().HasSynced,
+
+		excludeNamespaces: exclusionNSMap,
+	}
+
+	klog.Info("Setting up event handlers")
+	config.ServiceInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc: agentController.enqueueService,
+		UpdateFunc: func(old, newObj interface{}) {
+			agentController.handleUpdateService(old, newObj)
+		},
+		DeleteFunc: agentController.handleRemovedService,
+	}, handlerResync)
+
+	return agentController, nil
+}
+
+func (a *Controller) Run(stopCh <-chan struct{}) error {
+	var wg sync.WaitGroup
+	wg.Add(1)
+	defer utilruntime.HandleCrash()
+
+	// Start the informer factories to begin populating the informer caches
+	klog.Info("Starting IPAM Controller")
+
+	// Wait for the caches to be synced before starting workers
+	klog.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, a.servicesSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	go wait.Until(a.runServiceWorker, time.Second, stopCh)
+	wg.Wait()
```

Comment:
What is the reason for the `WaitGroup`? `Done` is never called so this will wait forever.

Reply:
Reusing this pattern from what I used in globalnet. Still find this a bit confusing.

Reply:
yeah - the `WaitGroup` here is not needed as you wait on `stopCh` below.  But it'll never get past this line 😄

Reply:
Now that I got basics working, will rework it. I still don't fully understand, so would help if you keep a sharp eye on this one :)